File: exam2/src/python3/graph.py
from typing import Dict, List, Optional

class Vertex:
    '''
    A vertex in a graph.
    
    Attributes:
        name (str): The name of the vertex.
        color (str): The color of the vertex.
        neighbors (List(vertex)): A list of the vertices that are a
          destination of an edge starting at this vertex.
        
    Methods:
        add_edge: Adds an edge to this vertex.
    '''
    name : str
    color : str
    neighbors : List["Vertex"]

    def __init__(self, name: str, color: str):
        '''
        Initializes a vertex.
        
        Arguments:
            name (str): The name of the vertex.
            color (str): The color of the vertex.
        '''
        self.name = name
        self.color = color
        self.neighbors = []

    def add_edge(self, dest: "Vertex") -> None:
        '''
        Adds an edge to this vertex.

        Arguments:
            dest (Vertex): The vertex to add an edge to.

        Returns:
            None
        '''
        self.neighbors.append(dest)


class Graph:
    '''
    A graph.

    Attributes:
        vertices (Dict[str, Vertex]): A dictionary of vertices in the graph.
            Note that these are vertex names mapped to vertex objects, and not
            vertex objects mapped to vertex objects representing edges.

    Methods:
        add_vertex: Adds a vertex to the graph.
        get_vertex: Gets a vertex from the graph.
    '''
    vertices : Dict[str, Vertex]

    def __init__(self):
        '''
        Initializes a graph.
        '''
        self.vertices = {}

    def add_vertex(self, vertex: Vertex) -> None:
        '''
        Adds a vertex to the graph.

        Arguments:
            vertex (Vertex): The vertex to add to the graph.

        Returns:
            None
        '''
        self.vertices[vertex.name] = vertex

    def get_vertex(self, name: str) -> Optional[Vertex]:
        '''
        Gets a vertex from the graph.

        Arguments:
            name (str): The name of the vertex to get.

        Returns:
            Optional[Vertex]: The vertex with the given name, or None if no
                vertex with that name exists.
        '''
        return self.vertices.get(name)

    

File: exam2/src/java/Graph.java
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

/**
 * A class for representing graphs
 *
 * Public attributes:
 *   vertices: the vertices in the graph represented as a map that
 *     maps vertex names to instances of the Vertex class
 *
 * Public methods:
 *   add_vertex: Adds a vertex to the graph.
 *   get_vertex: Gets a vertex from the graph.
 **/

public class Graph {
    public Map<String, Vertex> vertices;

    public Graph() {
        vertices = new HashMap<String, Vertex>();
    }

    /** Add vertex v to the graph
     *
     * Arguments:
     *   v: the vertex to add
     **/
    public void addVertex(Vertex v) {
        this.vertices.put(v.name, v);
    }

    /**
     * Get the vertex associated with a name
     *
     * Arguments:
     *   String name;
     *
     * Returns: the Vertex associated with the name or null
     **/
    public Vertex getVertex(String name) {
        if (this.vertices.containsKey(name)) {
            return this.vertices.get(name);
        } else {
            return null;
        }
    }

    public String toString() {
        String rv = "";
        List<String> vertexNames = new ArrayList<String>(this.vertices.keySet());
        Collections.sort(vertexNames);
        for (String vname : vertexNames) {
            Vertex v = this.getVertex(vname);
            rv += v + "\n";
        }
        return rv;
    }

}






    
